### ISSUE TITLE
need legally unencumbered vst SDK

### ISSUE THREAD

**Issue by zeromus** [zeromus/openvi#1]
Hello everyone! It seems steinberg has sent takedowns to a bunch of us hosting vst 2.x SDKs. Yeah, we should not have done that in the first place. I hope you don't mind that I've invited you all to discuss the matter. I think maybe if we get together we can make a more organized response -- or at least help some of us save time researching this all themselves. I admit I don't know much about all this--havent even started my big vsti project yet--but I put a couple of hours into studying the situation today which I wanted to share with everyone who got the same email I did.

It seems clear to me we need an open alternative to the vst SDK.  I really doubt the vst SDK is copyrightable as such, given that it's such a tiny bit of header files, but I think it's best not to fight that.

I would think the most straightforward way to proceed is for someone who's actively hacking on their plugins now to start building a new header file from scratch, until their plugins build once more. 

Please note, a new header file built that way is tainted by knowledge gleaned from the privileged SDK headers. Either you've seen the privileged headers or you've seen something coded by seeing the privileged headers. Personally, (I am not a lawyer) I think this not super important. Copyrightwise, rearranging their symbols into a new document should be fine, as long as it's sufficiently jumbled. The vst SDK headers do include some proprietary "creative" concepts for portable building which aren't at all required verbatim for effectively building vst components. These could be easily left out.

Some other people think it's important, and have worked on this as part of "VeSTige" with clean-room principles: https://github.com/LMMS/lmms/blob/master/include/aeffectx.h

It would take reworking your plugins in order to use this, and I'm not sure whether it's complete enough. However, even if it takes reworking--that may be for the best. It could be a chance to make a clean break from the tainted history.

Here's an example of someone recently incorporating "VeSTige" parts along with an incantation about "legally reverse engineered" to ward off legal demons: https://github.com/Ardour/ardour/commit/28f8772c3a75e051a9ed14edc4e5a888387dbe8b#diff-06bd75af8b863b1e6a5aa0439b8bffea 

In this case, it's thought he received a takedown strictly based on the filename, since at that time the contents of the file weren't even the vst SDK (simply named that way). I hope a couple of these folks won't mind me mentioning them in case they want to chime in: @pauldavisthefirst, @x42

Unfortunately the VeSTige header is GPL'd, most likely because it wasn't intended for this purpose; perhaps the contributors would let that go for the sake of interoperability? 

A compromise between the two aforementioned approaches would be doing a non-clean-room reinterpretation of the VST SDK with all new symbols. Anyone using that would be tainted with n-th degree-removed use of privileged information, but the copyright status would be much better with no shared literal text, qualms removed to a very abstruse realm where it's questionable whether the spirit of the API can be copyrighted. Note that this might not stop a lawyer from shooting at it if they knew where to aim, but it might render it invisible to the lawyers in question

Discussion questions:
* Does anyone else know of any other ready-to-go alternatives for the vst sdk?
* Does anyone know of any comparable situations with precedent?
* What do we know about reaper's vsti extensions? Do they have their own headers they have or could release in a format suitable for use by everyone?
* Were there other big batches of takedowns (probably?) Do you know anyone else affected who's figured out what how they're responding yet?
* Does anyone know what agreements may have been in effect years ago (entered into by joining the 'developer program' and downloading the files) before the SDK header files were changed to be slightly more legally sophisticated?
* Does anyone know enough about the law to opine on whether "this SDK must not be used to re-engineer or manipulate any technology used in any Steinberg or Third-party application or software module" includes what can only be fairly described as "engineer a third-party application or software module"? It seems their language was designed to address hacking, but I don't know what "re-engineer" means; and anyway it seems clear their business plan is to attempt to extinguish vsti 2.x, so shouldn't we assume that's what they meant to do with this text?
* What does "here are licensing agreements coupled with the use of development tools which you have to agree" mean when there are no licensing agreements provided? Note that there isn't a registration wall at this point. Meaning, effectively, right now, the 2.x SDK is provided without stipulations besides what's written in the headers. Is this a mistake?
* Can we compile a list of features missing from VeSTige, along with plugins using those features and subject matter experts for each (but _not_ the plugin authors) 
* Do the VeSTige authors have feedback as to how forked headers should be named? We probably shouldn't call them "VeSTige headers", that will get confusing.

Finally, here's everyone in the takedown batch I was in:

@mdsib @414n @8BitPimp @aardvarkk @AdrianGin @akionux @Alexey55 @ArTourter @askeksa @BaniMontoya @belaa007 @BenitoJedai @brucelane @Chameleon2525 @chock-mostlyharmless @chrisdonahue @clvn @cowyn @csound @cwilling @Deanmv @deepakgopinath @dmtaudio @dreieier @dunyakirkali @dxl5161 @e7mac @enzo1989 @eriser @faudio @forart @FranckFreiburger @FrankHartmanDSP @furesoft @garynewby @gmlion @godwincharan @gogins @gperon @grammaton @grimtraveller @hirano @impr0t @jakub-kudela @jamesehearn @jeffopolis @jnetterf @joshofreeness

**Comment by zeromus** [comment]
@kdsmoped @kode54 @konstantinkolesnik @krvajal @kxproject @laixinyuan @lauravelikonja @levaidaniel @lewisje @lfthomaz @maciejpiechoczek @madamdata @manchuwook @marianboda @martinbspheroid @massimobernava @mathieumailfert @mbrucher @mdeanda @metakine @mmi504 @mrotondo @mshavens @mvalente @nebyu @neobonzi @nickgang @ningmenglive @nyashkoshkko @obiwanjacobi @oliviermohsen @onozuka @osoumen @pac-dev @panosmdma @partikkelaudio @pepperpepperpepper @philkroos @phrb @pink-mist @piyush0011 @ponce @poulenc1899 @qwertyflame @raizor @rcgilbert @rchen1876 @reedham @remymuller @robjellinghaus

**Comment by zeromus** [comment]
@rockinroyle @romankubiak @sa66ath @sa-tsuklog @sethhochberg @sgrady2 @slackbuildsorg @slewrate @sound1ab @spencersalazar @square-waves @tanpitt @thanoscebolitas @theconservativecoder @theunreplicated @tiwb @tmp0910 @topher6345 @vjaksic @w41l @wakewakame @willysr @wiseqube @xia-lixun @yanmingsohu @yoyz @zephyrer @zonrobin @zuoteng @zyamusic @asnivor

**Comment by pauldavisthefirst** [comment]
I did not receive a takedown notice. I acted pre-emptively based on knowing someone else who did.

**Comment by x42** [comment]
> Does anyone know of any comparable situations with precedent?

I have received a takedown notice  last week https://github.com/github/dmca/blob/master/2018/2018-06-04-Steinberg-2.md for a project using the VeSTige header -- file identical to https://github.com/LMMS/lmms/blob/master/include/aeffectx.h  (!)

Since neither Ardour nor LMMS received the notice, I assume Steinberg goes only after plugins, not hosts.  That notion is also confirmed by https://www.steinberg.net/en/newsandevents/news/newsdetail/article/vst-2-coming-to-an-end-4727.html (quote: "VST 2 compatibility with Steinberg VST hosts will remain").

I assume they did simply search for "aeffectx.h" and hit me as false positive. I have meanwhile clarified that the header is "vestige", re-named the file and have not yet heard back from Steinberg. The project is still up on github.

I never agreed to Steinberg's SDK license personally, so I'm not directly affected either. Still I have meanwhile added some git mirrors besides github; most notably gitlab.com which is hosted in the EU (Netherlands), DCMA does not apply there,  APIs cannot be (C) in Europe and reverse engineering is legal.

**Comment by gogins** [comment]
This was a simple oversight on my part. I did not want to host any SDK
files but an earlier developer copied one of the header  files under a new
name and I didn't pay enough attention. The build system for my project
actually downloads the Steinberg SDK and uses that. I am assuming this is
OK. But I depend on the VST2 part of the SDK. Will this continue to be
available?

Thanks,
Mike

**Comment by gogins** [comment]
Oh, and the vst4cs opcodes for Csound ARE a host.

**Comment by yoyz** [comment]
I have remove aeffectx.h from the history of my repo too.
Seem, it is the file which was tracked down.

$ git filter-branch --force --index-filter 'git rm --cached --ignore-unmatch synthengine/vopm/dont_compile/aeffectx.h' --prune-empty --tag-name-filter cat -- --all                



 

    Le mercredi 13 juin 2018 à 22:44:29 UTC+2, Michael Gogins <notifications@github.com> a écrit :  
 
 Oh, and the vst4cs opcodes for Csound ARE a host.

**Comment by vjaksic** [comment]
I don't get it what is the big deal... why is it so important to them to
take down those headers... is there like nothing better to do...

**Comment by yoyz** [comment]
This header are baaaaaadd you knooooow ???
Dooon't use thooooose heaaaaderrrrr...
Seriously, I think it's like a vacum robot cleaner
 

    Le mercredi 13 juin 2018 à 22:51:53 UTC+2, Vladimir Jaksic <notifications@github.com> a écrit :  
 
 I don't get it what is the big deal... why is it so important to them to
take down those headers... is there like nothing better to do...

**Comment by yoyz** [comment]
Now I need to remove the whole directory targeted....
Hummmm, ok...
 

    Le mercredi 13 juin 2018 à 22:56:24 UTC+2, yoyz <yoyz2k@yahoo.fr> a écrit :  
 
 This header are baaaaaadd you knooooow ???
Dooon't use thooooose heaaaaderrrrr...
Seriously, I think it's like a vacum robot cleaner
 

    Le mercredi 13 juin 2018 à 22:51:53 UTC+2, Vladimir Jaksic <notifications@github.com> a écrit :  
 
 I don't get it what is the big deal... why is it so important to them to
take down those headers... is there like nothing better to do...

**Comment by x42** [comment]
> I have remove aeffectx.h from the history of my repo too. Seem, it is the file which was tracked down. $ git filter-branch --force --index-filter 'git rm --cached --ignore-unmatch synthengine/vopm/dont_compile/aeffectx.h' --prune-empty --tag-name-filter cat -- --all       

Was this the vestige header, named `aeffectx.h` or the actual Steinberg SDK?

For a project with a long clean history, like Ardour, I'd refuse to rewrite the 10+ year history and git force push invalidating all branches, especially since Ardour never incorporated the Steinberg SDK.

**Comment by pauldavisthefirst** [comment]
It seems reasonably clear that Steinberg's intention is to put an end to any new development that uses the VST2 API. They accept that they cannot stop people who already signed their license agreement from continuing to use the VST2 SDK to develop plugins (and hosts). But they want to stop new plugins (and hosts) from being developed using VST2 by new developers.

**Comment by zeromus** [comment]
Oddly, the website currently says "Please note that there are licensing agreements coupled with the use of development tools which you have to agree. These agreements contain conditions and/or requirements which are mostly concerned with liability issues. Please send the filled license agreement to the address that is given on the form, by mail or email it to info[SB]steinberg.de"

The VST 3.x SDK (containing 2.X) is then provided without registering, and it contains _no licensing agreements_ for VST 2.x. 

Several sources link to http://www.steinberg.net/en/company/developers.html as if the license would be there, but I can't find it.

The (revised at some point) 2.x headers do prohibit redistribution, as well as use to `re-engineer or manipulate any technology used in any Steinberg or Third-party application or software module` which does not sound to be as a prohibition of honest new development. But it does sound like a prohibition from referencing the SDK to re-engineer the API, as that is a "technology" used in both steinberg and third-party modules. 

And, I assume there was a similar restriction when signing up for the "developer program" historically.

Now for one other fun tidbit. The ASIO SDK _does_ come with an agreement, very much the kind you would expect to fill out and send to steinberg, much as the developer website says. Did they leave it out of the VST3 SDK archive? Supporting materials inside the VST3 SDK seem to suggest so. **But now for the actual tidbit**. The agreement in the ASIO SDK says "the agreement shall run for an unlimited period" and "In case that steinberg publishes a new version of the SDK according to a separate licensing agreement, steinberg is entitled to terminate _this_ agreement and any foregoing ASIO driver SDK licensing agreement regarding previous versions of the ASIO driver SDK with a 6 months written notice"

In other words, with 6 months notice they can tell you to stop using their SDK. If this is the same as the materials in the (missing) VST 2.x SDK, then even existing developers won't be able to (build? update? continue to sell? possess the SDK? whatever?) their old VST2 plugin products.

**Comment by pauldavisthefirst** [comment]
the VST2 SDK agreement is included in the VST2 SDK, and referenced in the
VST2 SDK, which is distinct (currently) from the VST3 SDK in the download
from Steinberg.

the prohibition in the headers is as old as VST itself, and was intended to
prevent (1) redistribution (2) reverse-engineering/duplication - basically,
Steinberg didn't care who used it, but they wanted to retain total control
of the VST API/specification. They said as much at the first meeting to
define GMPI, an attempt at a cross-industry audio/MIDI plugin API which
died (thanks to the MMA). They would not allow a cross-industry group to
take "control" of the VST API (if, for example, the GMPI working group (of
which I was a member) had decided that "VST was it").

in the 2000's everybody and their aunt was afraid of their APIs being taken
over by some third party which would then become the de facto standard. and
they were right to be worried ... reaper stepped right in and improved VST2
right away. but the redistribution clause stopped them from redistributing
the modified header(s).

Steinberg has already acknowledged online that they CANNOT stop people who
have already signed the VST2 SDK agreement from continuing to use and
develop with the VST2 SDK. Yvan Grabit has said this quite clearly. He has
also made it equally clear that they want to do everything they can to stop
anyone else doing anything with the VST2 SDK after October of this year.

**Comment by kode54** [comment]
I implement a host for MIDI playback, as a user mode driver that can load several VST 2 plugins that I doubt would work in VST 3 hosts. I deleted my repository from GitHub. I already have the repository mirrored elsewhere, Steinberg can suck it.

**Comment by brucelane** [comment]
I deleted my repository too...

**Comment by kode54** [comment]
I also kindly told the suppor@ address to tell the reporter to get bent, so that felt good.

I also feel good knowing that my macOS software will never have to touch Steinberg APIs to do anything with plugins, since the OS comes with handy APIs for plugins that don't require proprietary SDK licenses.

**Comment by x42** [comment]
>[mac] OS comes with handy APIs for plugins that don't require proprietary SDK licenses.

The Audio Unit SDK is likewise proprietary and not distributable. You cannot legally separate it from the OS. As opposed to Steinberg, Apple is in complete control and has forced annoying changes in the past.

> I deleted my repository too...

Isn't that a bit overkill?

**Comment by zeromus** [comment]
@pauldavisthefirst, the headers have not always had that prohibition.

for instance, see this old version that yvan grabit amusingly hosts himself: http://ygrabit.steinberg.de/~ygrabit/public_html/vstgui/V2.0/include/audioeffectx.h

or see this newer old version which is easily found on google: https://sourceforge.net/p/sonicbirth/svn/25/tree/SonicBirth_v1/ThirdParty/vst/audioeffectx.h

However it is possible that one would not have had access to those files without agreeing not to redistribute them. That's why I've been asking what the old agreements were.

I like your story that steinberg has been very aware and careful of this since the early 2000s. It's a cool (if annoying to us) master plan. However I'm not sure all the facts back this up. The facts also somewhat back up a simpler story of "oops, now we regret this".
